tuya.cpp

```diff
@@ -154,37 +162,112 @@ void DeRestPluginPrivate::handleTuyaClusterIndication(const deCONZ::ApsDataIndic
+                data = data << 8;
+                data = data + dummy;
+            }
         }
         
         //To be more precise
-        dp_identifier = (uint8_t) (dp & 0xFF);
-        dp_type = (uint8_t) ((dp >> 8) & 0xFF);
+        dp_identifier = (quint8) (dp & 0xFF);
+        dp_type = (quint8) ((dp >> 8) & 0xFF);
 
         DBG_Printf(DBG_INFO, "Tuya debug 4 : Address 0x%016llX Payload %s\n" , ind.srcAddress().ext(), qPrintable(zclFrame.payload().toHex()));
         DBG_Printf(DBG_INFO, "Tuya debug 5 : Status: %d Transid: %d Dp: %d (0x%02X,0x%02X) Fn: %d Data %ld\n", status , transid , dp , dp_type, dp_identifier, fn , data);
         
-        //Sensor and light use same cluster, so need to make a choice for device that have both
-        if ((sensorNode) && (lightNode))
+        if (length > 4) //schedule command
+        {
+            
+            // Monday = 64, Tuesday = 32, Wednesday = 16, Thursday = 8, Friday = 4, Saturday = 2, Sunday = 1
+            // If you want your schedule to run only on workdays, the value would be W124. (64+32+16+8+4 = 124)
+            // The API specifies 3 numbers, so a schedule that runs on Monday would be W064.
+            // Workday = W124
+            // Not working day = W003
+            // Saturday = W002
+            // Sunday = W001
+            // All days = W127
+            
+            QString transitions;
+            
+            length = length / 3;
+            
+            if (zclFrame.payload().size() < (( length * 3) + 6) )
+            {
+                DBG_Printf(DBG_INFO, "Tuya : Schedule data error\n");
+                return;
+            }
+            
+            quint8 hour;
+            quint8 minut;
+            quint8 heatSetpoint;
+            
+            quint8 part = 1;
+            QList<int> listday;
+            
+            if (dp == 0x0070) //work days (6)
+            {
+                listday << 124;
+            }
+            else if (dp == 0x00071) // holiday = Not working day (6)
+            {
+                listday << 003;
+            }
+            else if (dp == 0x0065) // Moe thermostat W124 (4) + W002 (4) + W001 (4)
+            {
+                part = length / 3;
+                listday << 124 << 2 << 1;
+            }
+            
+            for (; part > 0; part--)
+            {
+                for (; length > 0; length--)
+                {
+                    stream >> hour;
+                    stream >> minut;
+                    stream >> heatSetpoint;
+
+                    transitions += QString("T%1:%2|%3")
+                        .arg(hour, 2, 10, QChar('0'))
+                        .arg(minut, 2, 10, QChar('0'))
+                        .arg(heatSetpoint);
+                }
+            }
+            
+            if (part > 0 && listday.size() >= static_cast<int>(part))
+            {
+                updateThermostatSchedule(sensorNode, listday.at(part - 1), transitions);
+            }
```

Comment:
The check is good now, but `updateThermostatSchedule()` will never be called because due the prior for loop part will always be decremented down to 0.

Reply:
I have to admit I don't understand the logic here. Is this supposed to be within the loop?

Reply:
Yep, there is something not finished here https://github.com/dresden-elektronik/deconz-rest-plugin/pull/3457#discussion_r523438527
Not enought time to test it, and as I don't want to broke something, I had let the old tested version.

Have corrected the code, schedule is not the more important ATM.